chemprop/models/ffn.py

```diff
@@ -0,0 +1,355 @@
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor],
+        constraints: torch.tensor,
+        bond_types: torch.tensor,
+    ) -> torch.tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+
+        if constraints is not None:
+            if self.ffn_type == "atom":
+                output_hidden = self.ffn(hidden)
+                output = self.ffn_readout(output_hidden)
+            elif self.ffn_type == "bond":
+                b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+                b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+                output_1 = self.ffn(b_hidden_1)
+                output_2 = self.ffn(b_hidden_2)
+                output_hidden = (output_1 + output_2) / 2
+                output = self.ffn_readout(output_hidden)
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+            weights = self.weights_readout(output_hidden)
+            constrained_output = []
+            for i, (start, size) in enumerate(scope):
+                if size == 0:
+                    continue
+                else:
+                    cur_weights = weights.narrow(0, start, size)
+                    cur_output = output.narrow(0, start, size)
+
+                    cur_weights = torch.nn.Softmax(dim=0)(cur_weights).reshape(-1,)
+
+                    cur_output_sum = cur_output[:, 0].sum()
+
+                    cur_output[:, 0] = cur_output[:, 0] + cur_weights * (constraints[i] - cur_output_sum)
```

Comment:
```suggestion
                    cur_output[:, 0] = cur_output[:, 0] + cur_weights * (constraints[i] - cur_output[:, 0].sum())
```
it's more readable as a one-liner